app/views/licences/_fields.html.erb

```diff
@@ -3,14 +3,13 @@
     <fieldset class="inputs">
       <%= render :partial => 'shared/common_edition_attributes', :locals => {:f => f} %>
 
-      <div class="form-group">
-        <span class="form-label">
-          <%= f.label :licence_identifier %>
-        </span>
-        <span class="form-wrapper">
-          <%= f.text_field :licence_identifier, disabled: @resource.locked_for_edits?, class: "input-md-3 form-control" %>
-        </span>
-      </div>
+      <%= render partial: 'shared/validated_text_field', locals: {
+        f: f,
+        resource: resource,
+        id: "licence_identifier",
+        label: "Licence identifier",
+        input_classes: "input-md-3",
+      } %>
```

Comment:
I'm not convinced this partial based approach is quite wise, it seems like you'll have to create a lot of partials and think of lots of potential options.

It might be better to do things with some helper methods.

You could have a helper for just adding the error block:

something like below? with perhaps an attribute that can be used as a selector - could be used pretty much anywhere.
```
def form_field_errors(errors)
  tag.ul(class: %w[help-block js-error]) do
    safe_join errors.map { |e| tag.li(e) }
  end
end
```

Could then have a helper function that goes further for the most generic cases - which yields the field rather than needing to create it.

```
def form_group(form, field_name, label: nil, help: nil) 
  tag.div(class: "form_group) do
    label = #...
    field = tag.span(class: "form-wrapper") { yield }
    errors = form_field_errors(form.object.errors[field_name])
    help = #...
    safe_join(label, field, errors, help)
  end
end
```
which, in this instance could be called like this:

```
<%= form_group(f, :licence_identifier) do %>
  <%= f.text_field :licence_identifier, disabled: @resource.locked_for_edits?, class: "input-md-3 form-control" %>
<% end %>
```

Not 100% sure this would work mind.

Reply:
Thanks @kevindew - I have refactored into helpers as per your suggestion.